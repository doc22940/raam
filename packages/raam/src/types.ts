import { ResponsiveValue } from "styled-system";

// Sourced from: @types/theme-ui
export type Assign<T, U> = {
  [P in keyof (T & U)]: P extends keyof T
    ? T[P]
    : P extends keyof U
    ? U[P]
    : never;
};

export type GapProp = {
  /**
   * The gutter between items.
   */
  gap?: ResponsiveValue<string | number>;
};

export { ColorProps, FlexboxProps, SpaceProps } from "styled-system";
export { SxStyleProp } from "theme-ui";
export { InterpolationWithTheme } from "@emotion/core";
