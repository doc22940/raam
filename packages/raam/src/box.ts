// Extended from `@theme-ui/components` Box
// Excludes `variant`, adds an opinionated `reset`.
import styled from "@emotion/styled";
import { css } from "@theme-ui/css";
import { createShouldForwardProp } from "@styled-system/should-forward-prop";
import space from "@styled-system/space";
import color from "@styled-system/color";
import {
  Assign,
  ColorProps,
  SpaceProps,
  InterpolationWithTheme,
  SxStyleProp,
} from "./types";

export type BoxOwnProps = {
  as?: React.ElementType;
  sx?: SxStyleProp;
  css?: InterpolationWithTheme<any>;
} & ColorProps &
  SpaceProps;

export type BoxProps = Assign<
  React.ComponentPropsWithoutRef<"div">,
  BoxOwnProps
>;

const shouldForwardProp = createShouldForwardProp([
  ...space.propNames,
  ...color.propNames,
]);

const sx = props => css(props.sx)(props.theme);
const base = props => css(props.__css)(props.theme);
const reset = props =>
  (typeof props.as === "string" &&
    {
      /**
       * Ensure any list-items are rendered without bullets.
       * Adding a zero-width space in the content to prevent VoiceOver disable.
       */
      li: {
        listStyleType: "none",
        "&:before": {
          position: "absolute",
          content: '"\\200B"',
        },
      },
    }[props.as]) ||
  {};

export const Box = styled("div", {
  shouldForwardProp,
})(
  {
    boxSizing: "border-box",
    margin: 0,
    minWidth: 0,
  },
  base,
  reset,
  space,
  color,
  sx,
  props => props.css
);
