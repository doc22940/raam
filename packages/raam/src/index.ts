/**
 * @file Automatically generated by barrelsby.
 */

export * from "./inline";
export * from "./stack";
export * from "./wrap";
