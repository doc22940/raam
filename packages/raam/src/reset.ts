import { SxStyleProp } from "theme-ui";

type ResetLi = {
  listStyleTypeNone: SxStyleProp;
};

export const li: ResetLi = {
  listStyleTypeNone: {
    listStyleType: "none",
    "&:before": {
      position: "absolute",
      content:
        '"\\200B"' /* Add zero-width space to prevent VoiceOver disable */,
    },
  },
};
