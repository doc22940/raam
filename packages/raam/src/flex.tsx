import * as React from "react";
import { Box, BoxProps } from "./box";
import { determineElement } from "./utils";
import { GapProp } from "./types";

const flexMargin = ({
  gap,
  negative = false,
}: {
  negative?: boolean;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
} & GapProp) => (theme: any) => {
  if (typeof gap === "undefined" || gap === null) {
    return null;
  }

  // @ts-ignore
  const themedGap = theme.space[gap] || gap;
  const sign = negative ? -1 : 1;

  return typeof themedGap === "number"
    ? `${(themedGap / 2) * sign}px`
    : `calc(${themedGap} / 2 * ${sign})`;
};

const responsiveFlexMargin = ({
  gap,
  negative = false,
}: {
  /* I can't be bothered to fix this right now */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  negative?: boolean;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
} & GapProp) => (theme: any) => {
  if (Array.isArray(gap)) {
    return gap.map(gapValue => flexMargin({ gap: gapValue, negative })(theme));
  }

  if (typeof gap === "number" || typeof gap === "string") {
    return [flexMargin({ gap: gap, negative })(theme)];
  }

  return null;
};

export type FlexProps = BoxProps & GapProp;

export const Flex = React.forwardRef<any, FlexProps>(
  (
    { as = "div", gap, children, flex, flexDirection, flexWrap, ...props },
    ref
  ) => {
    const wrap = flexWrap === "wrap" || flexWrap === "wrap-reverse";
    const column =
      flexDirection === "column" || flexDirection === "column-reverse";

    return (
      <Box
        ref={ref}
        as={as}
        __css={{
          display: "flex",
          flexDirection,
          flexWrap,
          ...(gap && {
            overflow: wrap && "hidden",
            overflowX: !wrap && !column && "auto",
            margin:
              wrap &&
              (theme => responsiveFlexMargin({ gap, negative: true })(theme)),
          }),
        }}
        {...props}
      >
        {React.Children.map(children, (child, i) => (
          <Box
            as={determineElement(as)}
            __css={{
              position: "relative",
              ...(gap && wrap
                ? { margin: theme => responsiveFlexMargin({ gap })(theme) }
                : i > 0 && {
                    [column ? "marginTop" : "marginLeft"]: gap,
                  }),
            }}
            flex={flex || (gap && !column && "0 0 auto")}
          >
            {child}
          </Box>
        ))}
      </Box>
    );
  }
);
