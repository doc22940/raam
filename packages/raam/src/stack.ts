/** @jsx jsx */
import { forwardRef } from "react";
import { jsx, Box, BoxProps, GridProps } from "theme-ui";
import { li } from "./reset";

export type StackProps = BoxProps & Pick<GridProps, "gap">;

// https://github.com/DefinitelyTyped/DefinitelyTyped/issues/35834#issuecomment-497445051
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const Stack = forwardRef<any, StackProps>(
  ({ gap = 3, sx, children, ...props }, ref) =>
    jsx(
      Box,
      {
        ...props,
        ref,
        sx: {
          /**
           * If a browser supports the `grid-gap` property, let's use it.
           * Otherwise, fallback to the lobotomized owl selector to style children.
           */
          "@supports (grid-gap: 0)": {
            /* Theme UI's types don't seem to like me doing this */
            // @ts-ignore
            display: "grid",
            // @ts-ignore
            gridGap: gap,
          },
          "@supports not (grid-gap: 0)": {
            "& > * + *": {
              marginTop: gap,
            },
          },
          /* Ensure direct child list-items render without bullets */
          "& > li": li.listStyleTypeNone,
          ...(sx && sx),
        },
      },
      children
    )
);
