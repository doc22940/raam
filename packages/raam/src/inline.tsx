import * as React from "react";
import { FlexProps, Flex } from "./flex";

export type InlineProps = FlexProps;

export const Inline = React.forwardRef<any, InlineProps>(
  (
    { gap = 3, alignItems = "center", flexDirection = "row", ...props },
    ref
  ) => (
    <Flex
      ref={ref}
      alignItems={alignItems}
      gap={gap}
      flexDirection={flexDirection}
      {...props}
    />
  )
);
