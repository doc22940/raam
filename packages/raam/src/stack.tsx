import * as React from "react";
import { FlexProps, Flex } from "./flex";

export type StackProps = FlexProps;

export const Stack = React.forwardRef<any, StackProps>(
  ({ gap = 3, flexDirection = "column", ...props }, ref) => (
    <Flex ref={ref} gap={gap} flexDirection={flexDirection} {...props} />
  )
);
