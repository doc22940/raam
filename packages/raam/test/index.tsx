import * as React from "react";
import renderer from "react-test-renderer";
import { matchers } from "jest-emotion";
import { Flex } from "../src";

expect.extend(matchers);

const render = (el: React.ReactNode) => renderer.create(el);
const renderJSON = (el: React.ReactNode) => render(el).toJSON();

describe("Flex", () => {
  test("renders with an array of children", () => {
    const json = renderJSON(<Flex>{["Item 1", "Item 2", "Item 3"]}</Flex>);
    // expect(json).toHaveStyleRule("padding", "8px");
    expect(json).toMatchSnapshot();
  });

  test("renders as a list", () => {
    const ordered = renderJSON(
      <Flex as="ol">
        <a href="#1">Item 1</a>
        <a href="#2">Item 2</a>
        <a href="#3">Item 3</a>
      </Flex>
    );
    expect(ordered).toMatchSnapshot();

    const unordered = renderJSON(
      <Flex as="ul">
        <a href="#1">Item 1</a>
        <a href="#2">Item 2</a>
        <a href="#3">Item 3</a>
      </Flex>
    );
    expect(unordered).toMatchSnapshot();
  });
});
