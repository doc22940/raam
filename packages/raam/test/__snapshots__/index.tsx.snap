// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flex renders as a list 1`] = `
.emotion-3 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  position: relative;
  list-style-type: none;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.emotion-0:before {
  position: absolute;
  content: "\\200B";
}

<ol
  className="emotion-3"
>
  <li
    className="emotion-0"
  >
    <a
      href="#1"
    >
      Item 1
    </a>
  </li>
  <li
    className="emotion-0"
  >
    <a
      href="#2"
    >
      Item 2
    </a>
  </li>
  <li
    className="emotion-0"
  >
    <a
      href="#3"
    >
      Item 3
    </a>
  </li>
</ol>
`;

exports[`Flex renders as a list 2`] = `
.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  position: relative;
  list-style-type: none;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.emotion-0:before {
  position: absolute;
  content: "\\200B";
}

.emotion-3 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding-left: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

<ul
  className="emotion-3"
>
  <li
    className="emotion-0"
  >
    <a
      href="#1"
    >
      Item 1
    </a>
  </li>
  <li
    className="emotion-0"
  >
    <a
      href="#2"
    >
      Item 2
    </a>
  </li>
  <li
    className="emotion-0"
  >
    <a
      href="#3"
    >
      Item 3
    </a>
  </li>
</ul>
`;

exports[`Flex renders with an array of children 1`] = `
.emotion-3 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  position: relative;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

<div
  className="emotion-3"
>
  <div
    className="emotion-0"
  >
    Item 1
  </div>
  <div
    className="emotion-0"
  >
    Item 2
  </div>
  <div
    className="emotion-0"
  >
    Item 3
  </div>
</div>
`;
